middleware.ts

```diff
@@ -0,0 +1,37 @@
+import { withAuth } from "next-auth/middleware";
+// middleware.ts
+import { NextRequest, NextResponse } from "next/server";
+import { getToken } from "next-auth/jwt";
+
+export async function middleware(req: NextRequest) {
+  // Skip middleware for static files and API routes
+
+  const session = await getToken({ req, secret: process.env.NEXTAUTH_SECRET });
+  const { pathname } = req.nextUrl;
+
+  // Redirect to sign-up if there's no session and not on the sign-up page
+  if (!session && pathname !== "/sign-up") {
+    const url = req.nextUrl.clone();
+    url.pathname = "/sign-up";
+    return NextResponse.redirect(url);
+  }
+
+  return NextResponse.next();
+}
+
+export default withAuth({
+  callbacks: {
+    authorized({ req, token }: { req: NextRequest; token: any }) {
+      // `/admin` requires admin role
+      if (req.nextUrl.pathname === "/admin") {
+        return token?.userRole === "admin";
+      }
+      // `/me` only requires the user to be logged in
+      return !!token;
+    },
+  },
+});
+
+export const config = {
+  matcher: ["/((?!api|_next/static|_next/image|images|icons|favicon.ico).*)"],
```

Comment:
```suggestion
  matcher: ["/((?!api|_next/static|_next/image|sign-in|sign-up|favicon.ico).*)"],
```
I wouldn't recommend doing it this way, you can if you'd like but following the docs and doing local imports makes more sense. I would also just add sign-in and sign-up as routes that are skipped here, so there's no redirect when the user tries to sign-in.

You can also remove some logic because of this change.

Reply:
will get on that after I'm done with the layouts, thanks for the comments